unit_tests/request_client_server/test_main.cpp

```diff
@@ -0,0 +1,346 @@
+
+  init_globals ();
+
+  req_server.start_thread ();
+
+  send_request_id_as_message (req_client, reqids::_0);
+  send_request_id_as_message (req_client, reqids::_1);
+  send_request_id_as_message (req_client, reqids::_1);
+  send_request_id_as_message (req_client, reqids::_1);
+  send_request_id_as_message (req_client, reqids::_0);
+  send_request_id_as_message (req_client, reqids::_1);
+  send_request_id_as_message (req_client, reqids::_0);
+
+  sockdir.wait_for_all_messages ();
+
+  req_server.stop_thread ();
+
+  require_all_sent_requests_are_handled ();
+
+  clear_socket_directions ();
+}
+
+TEST_CASE ("Two client-server communicate with each-other", "")
+{
+  // Test how two request_client_server instances interact with each-other. The main thread of the test case acts as
+  // the client for both request_client_server and sends them requests. The requests are handled by the internal
+  // threads of each request_client_server.
+  //
+  // Verify that:
+  //    - all the requests, to both request_client_server instances, are handled by the expected type of handler
+  //    - the number of requests sent is same as the number of requests handled.
+  //
+
+  test_request_client_server_type_one req_client_server_1 = create_request_client_server_one ();
+  test_request_client_server_type_two req_client_server_2 = create_request_client_server_two ();
+
+  mock_socket_direction sockdir_1_to_2;
+  mock_socket_direction sockdir_2_to_1;
+  mock_socket_between_two_client_servers (req_client_server_1, req_client_server_2, sockdir_1_to_2, sockdir_2_to_1);
+
+  init_globals ();
+
+  req_client_server_1.start_thread ();
+  req_client_server_2.start_thread ();
+
+  // Mix sends on both ends
+  send_request_id_as_message (req_client_server_1, reqids_1_to_2::_0);
+  send_request_id_as_message (req_client_server_2, reqids_2_to_1::_0);
+  send_request_id_as_message (req_client_server_1, reqids_1_to_2::_1);
+  send_request_id_as_message (req_client_server_1, reqids_1_to_2::_1);
+  send_request_id_as_message (req_client_server_2, reqids_2_to_1::_0);
+  send_request_id_as_message (req_client_server_1, reqids_1_to_2::_0);
+  send_request_id_as_message (req_client_server_1, reqids_1_to_2::_0);
+  send_request_id_as_message (req_client_server_1, reqids_1_to_2::_1);
+  send_request_id_as_message (req_client_server_1, reqids_1_to_2::_1);
+  send_request_id_as_message (req_client_server_1, reqids_1_to_2::_1);
+  send_request_id_as_message (req_client_server_2, reqids_2_to_1::_1);
+  send_request_id_as_message (req_client_server_1, reqids_1_to_2::_1);
+  send_request_id_as_message (req_client_server_2, reqids_2_to_1::_0);
+  send_request_id_as_message (req_client_server_2, reqids_2_to_1::_1);
+  send_request_id_as_message (req_client_server_2, reqids_2_to_1::_0);
+  send_request_id_as_message (req_client_server_2, reqids_2_to_1::_1);
+  send_request_id_as_message (req_client_server_2, reqids_2_to_1::_0);
```

Comment:
For more robustness, would it make sense for all the messages be sent by a couple of threads on each side?

Reply:
Sending requests on request_client & request_client_server is not thread safe. Thread-safe sending is going to be treated separately in LETS-26.

Reply:
Agreed